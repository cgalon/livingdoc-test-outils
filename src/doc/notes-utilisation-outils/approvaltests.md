## ApprovalTests

### Description

Framework de test basé sur les résultats d'une API.
Permet d'enregistrer le comportement d'une API "en boite noire" puis de vérifier que les tests passent toujours suite à un refactoring.

### Site / docs complémentaires

[Site](https://approvaltests.com/)
[ApprovalTests.Java](https://github.com/approvals/ApprovalTests.Java)

### Intégration

Seulement deux dépendances à ajouter (Approval + sérialisation JSON) :
```xml
        <dependency>
            <groupId>com.approvaltests</groupId>
            <artifactId>approvaltests</artifactId>
            <version>18.6.0</version>
        </dependency>
```

### Commentaires

- Les tests sont joués par JUnit en même temps que les tests unitaires classiques.
- Fonctionnement basé sur la découverte du comportement du System Under Test
  - Le premier run d'un test produit un fichier de référence de ce qui est attendu en sortie.
  - On renomme le fichier de résultat produit pour "acter" le résultat attendu.
  - Les tests suivants vérifient le résultat obtenu avec le résultat attendu dans le fichier de référence.
- Les fichiers de référence sont produits automatiquement, mais nécessitent un renommage pour être pris en compte comme référence.
- Parait très intéressant pour tester et "acter" le comportement d'un SUT en boite noire puis pour le refactorer.
- C'est du "test after".
- En l'état, ne sert pas vraiment à documenter, mais à assurer de la non-regression pour de la maintenance ou du refactoring. 
