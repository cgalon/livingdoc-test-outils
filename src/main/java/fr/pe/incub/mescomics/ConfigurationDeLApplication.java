package fr.pe.incub.mescomics;

import io.swagger.v3.oas.models.OpenAPI;
import io.swagger.v3.oas.models.info.Contact;
import io.swagger.v3.oas.models.info.Info;
import org.springframework.boot.autoconfigure.domain.EntityScan;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.data.jpa.repository.config.EnableJpaRepositories;
import org.springframework.jdbc.datasource.DriverManagerDataSource;
import org.springframework.orm.jpa.JpaVendorAdapter;
import org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean;
import org.springframework.orm.jpa.vendor.HibernateJpaVendorAdapter;

import javax.persistence.EntityManager;
import javax.sql.DataSource;

@Configuration
@EntityScan(basePackageClasses = {Application.class})
@EnableJpaRepositories("fr.pe.incub.mescomics")
public class ConfigurationDeLApplication {

    @Bean
    public DataSource dataSource() {
        DriverManagerDataSource dataSource = new DriverManagerDataSource();
        dataSource.setDriverClassName("org.h2.Driver");
        dataSource.setUrl("jdbc:h2:mem:testdb");
        dataSource.setUsername("sa");
        dataSource.setPassword("");
        return dataSource;
    }

    @Bean
    public EntityManager entityManager() {
        return entityManagerFactory().getObject().createEntityManager();
    }

    @Bean
    public LocalContainerEntityManagerFactoryBean entityManagerFactory() {
        LocalContainerEntityManagerFactoryBean em = new LocalContainerEntityManagerFactoryBean();
        em.setDataSource(dataSource());
        em.setPackagesToScan("fr.pe.incub.mescomics");
        em.setJpaVendorAdapter(jpaVendorAdapter());
        return em;
    }

    @Bean
    public JpaVendorAdapter jpaVendorAdapter() {
        JpaVendorAdapter jva = new HibernateJpaVendorAdapter();
        return jva;
    }

    @Bean
    public OpenAPI configureLesMetadataDeLaDocumentation() {
        return new OpenAPI()
                .info(new Info()
                        .title("Gestion d'une collection de comics")
                        .description("Application d'exemple de génération de documentation")
                        .version("1.0.0")
                        .contact(new Contact()
                                .name("Christophe Galon")
                                .email("cgalon@free.fr")
                                .url("https://github.com/cgalon")));
    }

}
