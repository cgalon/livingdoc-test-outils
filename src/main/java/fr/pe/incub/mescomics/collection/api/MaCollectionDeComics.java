package fr.pe.incub.mescomics.collection.api;

import fr.pe.incub.mescomics.collection.domaine.Collection;
import fr.pe.incub.mescomics.collection.domaine.Comics;
import fr.pe.incub.mescomics.collection.domaine.ServiceDeCollection;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
public class MaCollectionDeComics {

    ServiceDeCollection serviceDeCollection;

    @Autowired
    public MaCollectionDeComics(ServiceDeCollection serviceDeCollection) {
        this.serviceDeCollection = serviceDeCollection;
    }

    public List<Comics> retourneMaListeDeComicsPourCeTitre(String titre) {
        Collection maCollection = serviceDeCollection.recupererMaCollectionCompleteDeComics();
        return maCollection.recupereLaListeDesComicsDeLaSerie(titre);
    }

}
