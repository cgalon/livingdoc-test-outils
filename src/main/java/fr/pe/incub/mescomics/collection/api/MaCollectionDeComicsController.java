package fr.pe.incub.mescomics.collection.api;

import fr.pe.incub.mescomics.collection.domaine.Comics;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

import java.util.List;

/**
 * Ressource de gestion de la collection de Comics détenus par l'utilisateur.
 */
@RestController
@RequestMapping("/collection/titre")
@Tag(name="Collection de comics", description="Ressource de gestion de la collection de Comics détenus par " +
        "l'utilisateur.")
public class MaCollectionDeComicsController {

    MaCollectionDeComics maCollectionDeComics;

    @Autowired
    public MaCollectionDeComicsController(@Autowired MaCollectionDeComics maCollectionDeComics) {
        this.maCollectionDeComics = maCollectionDeComics;
    }

    @Operation(description ="Retourne une liste de comics pour la titre passé en paramètre.")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Liste trouvée"),
            @ApiResponse(responseCode = "401", description = "Opération non autorisée"),
            @ApiResponse(responseCode = "403", description = "Opération interdite"),
            @ApiResponse(responseCode = "404", description = "Pas de comics trouvé pour ce titre")
    })
    @RequestMapping(method = RequestMethod.GET)
    public List<Comics> retourneMaListeDeComicsPourCeTitre(@RequestParam String titre) {
        return maCollectionDeComics.retourneMaListeDeComicsPourCeTitre(titre);
    }

    /**
     * Non implémenté. Existe juste pour tester la génération de documentation.
     * @param titre
     */
    @Operation(description ="Supprime la liste de comics correspondants au titre passé en paramètre.")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Suppression effectuée avec succès"),
            @ApiResponse(responseCode = "401", description = "Opération non autorisée"),
            @ApiResponse(responseCode = "403", description = "Opération interdite"),
            @ApiResponse(responseCode = "404", description = "Pas de comics à supprimer pour ce titre")
    })
    @RequestMapping(method = RequestMethod.DELETE)
    @ResponseStatus(value = HttpStatus.OK)
    public void supprimeUneListeDeComicsPourCeTitre(@RequestParam String titre) {

    }
}
