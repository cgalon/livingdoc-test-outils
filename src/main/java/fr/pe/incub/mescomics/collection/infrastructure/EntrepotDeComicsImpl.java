package fr.pe.incub.mescomics.collection.infrastructure;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import fr.pe.incub.mescomics.collection.domaine.Comics;
import fr.pe.incub.mescomics.collection.domaine.EntrepotDeComics;

@Component
public class EntrepotDeComicsImpl implements EntrepotDeComics {

    EntrepotDeComicsSpring entrepotDeComics;

    @Autowired
    public EntrepotDeComicsImpl(EntrepotDeComicsSpring entrepotDeComics) {
        this.entrepotDeComics = entrepotDeComics;
    }

    public Iterable<Comics> retrouveTousLesComics() {
        return entrepotDeComics.findAll();
    };

}
