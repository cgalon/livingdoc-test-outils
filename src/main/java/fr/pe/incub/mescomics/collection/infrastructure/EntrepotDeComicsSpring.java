package fr.pe.incub.mescomics.collection.infrastructure;

import fr.pe.incub.mescomics.collection.domaine.Comics;
import org.springframework.data.repository.CrudRepository;
import org.springframework.stereotype.Repository;

@Repository
public interface EntrepotDeComicsSpring extends CrudRepository<Comics, Integer> {

}
