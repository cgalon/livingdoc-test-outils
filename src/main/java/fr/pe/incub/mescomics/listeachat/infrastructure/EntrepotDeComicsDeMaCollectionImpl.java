package fr.pe.incub.mescomics.listeachat.infrastructure;

import fr.pe.incub.mescomics.collection.api.MaCollectionDeComics;
import fr.pe.incub.mescomics.collection.domaine.Comics;
import fr.pe.incub.mescomics.listeachat.domaine.EntrepotDeComicsDeMaCollection;
import fr.pe.incub.mescomics.listeachat.domaine.NumeroPresent;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;

@Component
public class EntrepotDeComicsDeMaCollectionImpl implements EntrepotDeComicsDeMaCollection{

    private MaCollectionDeComics maCollectionDeComics;

    @Autowired
    public EntrepotDeComicsDeMaCollectionImpl(MaCollectionDeComics maCollectionDeComics) {
        this.maCollectionDeComics = maCollectionDeComics;
    }

    public List<NumeroPresent> retrouveLesComicsQueJePossedePourCeTitre(String titre) {
        List<Comics> lesComicsDeMaCollection = maCollectionDeComics.retourneMaListeDeComicsPourCeTitre(titre);
        return lesComicsDeMaCollection.stream().map(transformeEnNumeroPresent).collect(Collectors.toList());
    }

    private Function<Comics, NumeroPresent> transformeEnNumeroPresent = comics -> {
        NumeroPresent numero = new NumeroPresent(comics.titreDeLaRevue, comics.numero);
        return numero;
    };
}
