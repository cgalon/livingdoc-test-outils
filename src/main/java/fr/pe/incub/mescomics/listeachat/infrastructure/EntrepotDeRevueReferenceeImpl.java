package fr.pe.incub.mescomics.listeachat.infrastructure;

import fr.pe.incub.mescomics.listeachat.domaine.EntrepotDeRevueReferencee;
import fr.pe.incub.mescomics.listeachat.domaine.RevueReferencee;
import fr.pe.incub.mescomics.referentiel.api.ReferentielDeRevues;
import fr.pe.incub.mescomics.referentiel.domaine.Numero;
import fr.pe.incub.mescomics.referentiel.domaine.Revue;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.Optional;


@Service
public class EntrepotDeRevueReferenceeImpl implements EntrepotDeRevueReferencee {

    ReferentielDeRevues referentielDeRevues;

    @Autowired
    public EntrepotDeRevueReferenceeImpl(ReferentielDeRevues referentielDeRevues) {
        this.referentielDeRevues = referentielDeRevues;
    }

    public RevueReferencee retrouveLaRevuePourCeTitre(String titre) {

        RevueReferencee revueTrouvee = null;
        Optional<Revue> revue = referentielDeRevues.retrouveLaRevuePourCeTitre(titre);
        if (revue.isPresent()) {
            Optional<Numero> premierNumero = referentielDeRevues.retrouveLePremierNumeroPourCetteRevue(revue.get());
            if (premierNumero.isPresent()) {
                revueTrouvee = new RevueReferencee(revue.get().titre, revue.get().recupereLeNombreDeNumeros(), premierNumero.get().dateDeParution, revue.get().nomDeLEditeur);
            }
        }
        return revueTrouvee;
    }
}
