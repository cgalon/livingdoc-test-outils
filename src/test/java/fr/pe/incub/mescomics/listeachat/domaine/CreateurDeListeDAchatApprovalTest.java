package fr.pe.incub.mescomics.listeachat.domaine;

import fr.pe.incub.mescomics.UtilitairesDeTest;
import fr.pe.incub.mescomics.listeachat.infrastructure.EntrepotDeComicsDeMaCollectionImpl;
import fr.pe.incub.mescomics.listeachat.infrastructure.EntrepotDeRevueReferenceeImpl;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.when;

/**
 * Utilisation du framework Approvals.
 * Reprise des tests de la classe CreateurDeListeDAchatTest.
 * Les "Arrange" et "Act" de chaque test sont les mêmes. Seuls les "Assert" sont modifiés pour utiliser ApprovalTests.
 */
@ExtendWith(MockitoExtension.class)
public class CreateurDeListeDAchatApprovalTest {

    @Mock
    private EntrepotDeComicsDeMaCollectionImpl recuperateurDeMaCollectionDeComicsBouchon;

    @Mock
    private EntrepotDeRevueReferenceeImpl recuperateurDeRevueReferenceeBouchon;

    UtilitairesDeTest utilitairesDeTest = new UtilitairesDeTest();

    @Test
    public void doitCreerUneListeDAchatVideLorsqueJAiTousLesNumeros() {
        when(recuperateurDeMaCollectionDeComicsBouchon.retrouveLesComicsQueJePossedePourCeTitre("Spidey")).thenReturn(utilitairesDeTest.creeUneListeCompleteDeNumerosPresentsDeSpidey());
        when(recuperateurDeRevueReferenceeBouchon.retrouveLaRevuePourCeTitre("Spidey")).thenReturn(utilitairesDeTest.creeUneRevueReferenceePourSpidey());
        CreateurDeListeDAchat createurDeListeDAchat = new CreateurDeListeDAchat(recuperateurDeMaCollectionDeComicsBouchon, recuperateurDeRevueReferenceeBouchon);

        ListeDAchat listeDAchatSpidey = createurDeListeDAchat.creeUneListeDAchatPourUneRevue("Spidey");

        Approvals.verify(listeDAchatSpidey.nombreDeNumerosManquants());
    }

    @Test
    public void doitCreerUneListeDAchatCompleteLorsqueJeNAiAucunNumero() {
        when(recuperateurDeMaCollectionDeComicsBouchon.retrouveLesComicsQueJePossedePourCeTitre("Titan")).thenReturn(utilitairesDeTest.creeUneListeDeNumerosPresentsVide());
        when(recuperateurDeRevueReferenceeBouchon.retrouveLaRevuePourCeTitre("Titan")).thenReturn(utilitairesDeTest.creeUneRevueReferenceePourTitan());
        CreateurDeListeDAchat createurDeListeDAchat = new CreateurDeListeDAchat(recuperateurDeMaCollectionDeComicsBouchon, recuperateurDeRevueReferenceeBouchon);

        ListeDAchat listeDAchatTitan = createurDeListeDAchat.creeUneListeDAchatPourUneRevue("Titan");

        Approvals.verifyAll("Numéro de Titan manquant", listeDAchatTitan.listeDesNumerosManquants());
    }

    @Test
    public void doitCreerUneListeDAchatAPartirDUneCollectionExistante() {
        when(recuperateurDeMaCollectionDeComicsBouchon.retrouveLesComicsQueJePossedePourCeTitre("Strange")).thenReturn(utilitairesDeTest.creeUneListeIncompleteDeNumerosPresentsDeStrange());
        when(recuperateurDeRevueReferenceeBouchon.retrouveLaRevuePourCeTitre("Strange")).thenReturn(utilitairesDeTest.creeUneRevueReferenceePourStrange());
        CreateurDeListeDAchat createurDeListeDAchat = new CreateurDeListeDAchat(recuperateurDeMaCollectionDeComicsBouchon, recuperateurDeRevueReferenceeBouchon);

        ListeDAchat listeDAchatStrange = createurDeListeDAchat.creeUneListeDAchatPourUneRevue("Strange");

        Approvals.verifyAll("Numéro de Strange manquant", listeDAchatStrange.listeDesNumerosManquants());
    }

}