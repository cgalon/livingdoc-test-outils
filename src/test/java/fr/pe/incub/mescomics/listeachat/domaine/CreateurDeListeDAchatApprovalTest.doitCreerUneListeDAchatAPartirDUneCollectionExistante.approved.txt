Numéro de Strange manquant[00] = NumeroManquant{serie='Strange', numero=2}
Numéro de Strange manquant[01] = NumeroManquant{serie='Strange', numero=3}
Numéro de Strange manquant[02] = NumeroManquant{serie='Strange', numero=4}
Numéro de Strange manquant[03] = NumeroManquant{serie='Strange', numero=6}
Numéro de Strange manquant[04] = NumeroManquant{serie='Strange', numero=7}
Numéro de Strange manquant[05] = NumeroManquant{serie='Strange', numero=8}
Numéro de Strange manquant[06] = NumeroManquant{serie='Strange', numero=9}
Numéro de Strange manquant[07] = NumeroManquant{serie='Strange', numero=13}
Numéro de Strange manquant[08] = NumeroManquant{serie='Strange', numero=14}
Numéro de Strange manquant[09] = NumeroManquant{serie='Strange', numero=16}
Numéro de Strange manquant[10] = NumeroManquant{serie='Strange', numero=17}
Numéro de Strange manquant[11] = NumeroManquant{serie='Strange', numero=18}
Numéro de Strange manquant[12] = NumeroManquant{serie='Strange', numero=19}
Numéro de Strange manquant[13] = NumeroManquant{serie='Strange', numero=20}
